itest/rust/src/builtin_tests/containers/signal_test.rs

```diff
@@ -776,3 +776,397 @@ mod custom_callable {
+        });
+    }
+
+    #[itest]
+    fn no_panic_when_disconnecting_on_drop(ctx: &TestContext) {
+        const NUM_NODES: i32 = 10;
+
+        let mut tree = ctx.scene_tree.clone();
+
+        let mut broadcaster = SignalNode::new_alloc();
+        tree.add_child(&broadcaster);
+
+        // Spawn and connect all
+        for i in 0..NUM_NODES {
+            let mut safe_node = DisconnectOnDropNode::new_alloc();
+            broadcaster.add_child(&safe_node);
+
+            let connection_handle = broadcaster
+                .signals()
+                .notify()
+                .builder()
+                .connect_other_mut(&safe_node, DisconnectOnDropNode::increment_self);
+
+            let mut bind = safe_node.bind_mut();
+            bind.set_counter(i);
+            bind.connections_to_me.push(connection_handle);
+            drop(bind);
+        }
+
+        // Emitting should work as intended
+        broadcaster.signals().notify().emit();
+
+        // Free some of the nodes
+        assert!(broadcaster.get_child_count() == NUM_NODES);
+        for (i, child) in broadcaster.get_children().iter_shared().enumerate() {
+            let should_free = i % 3 == 0;
+            if should_free {
+                child.free();
+            }
+        }
+
+        // Since connections are disonnected on drop this should still work
+        broadcaster.signals().notify().emit();
+
+        broadcaster.free();
+    }
+
+    #[should_panic]
+    #[itest]
+    fn double_disconnect_panic(ctx: &TestContext) {
+        // Handle should panic if we disconnect something that is not connected
+        connection_handles::disconnect_potentially_already_disconnected(ctx, false);
+    }
+
+    #[itest]
+    fn double_disconnect_with_check(ctx: &TestContext) {
+        // Handle should NOT panic if we check connection integrity first
+        connection_handles::disconnect_potentially_already_disconnected(ctx, true);
+    }
+
+    // Helper functions:
+
+    fn disconnect_self_builder(
+        ctx: &TestContext,
+        handle_function: impl FnOnce(&Gd<SignalNode>) -> ConnectHandle,
+    ) {
+        let mut tree = ctx.scene_tree.clone();
+
+        let notifier = SignalNode::new_alloc();
+        tree.add_child(&notifier);
+
+        let connection_handle = handle_function(&notifier);
+
+        assert!(!notifier
+            .get_signal_connection_list(SignalNode::NOTIFY_NAME)
+            .is_empty());
+
+        assert_eq!(notifier.bind().get_counter(), 0i32);
+
+        // Emit signal
+        notifier.signals().notify().emit();
+
+        assert_eq!(notifier.bind().get_counter(), 1i32);
+
+        // Disconnect
+        connection_handle.disconnect();
+
+        assert!(notifier
+            .get_signal_connection_list(SignalNode::NOTIFY_NAME)
+            .is_empty());
+
+        // Emit signal after disconnect
+        notifier.signals().notify().emit();
+
+        assert_eq!(notifier.bind().get_counter(), 1i32);
+
+        notifier.free();
```

Comment:
An empty line between _all_ statements makes code very hard to read. Empty lines are good, but please groups of related statements, such as declarations, asserts, etc.

Also, comments with . :slightly_smiling_face: